src/whir/prover.rs

```diff
@@ -390,21 +414,43 @@ where
             folding_factor,
             self.0.final_queries,
             prover_state,
+            &self.0.deduplication_strategy,
         )?;
 
-        let merkle_proof = round_state
-            .prev_merkle
-            .generate_multi_proof(final_challenge_indexes.clone())
-            .unwrap();
         // Every query requires opening these many in the previous Merkle tree
         let fold_size = 1 << folding_factor;
         let answers = final_challenge_indexes
-            .into_iter()
+            .iter()
             .map(|i| round_state.prev_merkle_answers[i * fold_size..(i + 1) * fold_size].to_vec())
             .collect::<Vec<_>>();
 
         prover_state.hint::<Vec<Vec<F>>>(&answers)?;
-        prover_state.hint::<MultiPath<MerkleConfig>>(&merkle_proof)?;
+
+        match self.0.merkle_proof_strategy {
```

Comment:
Let's leave a note to eventually refactor this to a Merkle module (similar to the other utils.rs)